{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_FOLDER = 'Data' # Use the data folder provided in Tutorial 02 - Intro to Pandas.\n",
    "\n",
    "import pandas as pd\n",
    "import glob\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Daily average new cases</th>\n",
       "      <th>Daily average deaths</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Country</th>\n",
       "      <th>Month</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Guinea</th>\n",
       "      <th>2014-8</th>\n",
       "      <td>4.161290</td>\n",
       "      <td>0.548387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-9</th>\n",
       "      <td>10.466667</td>\n",
       "      <td>1.900000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-10</th>\n",
       "      <td>1.096774</td>\n",
       "      <td>0.483871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">Liberia</th>\n",
       "      <th>2014-6</th>\n",
       "      <td>1.333333</td>\n",
       "      <td>0.466667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-7</th>\n",
       "      <td>3.032258</td>\n",
       "      <td>1.516129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-8</th>\n",
       "      <td>10.806452</td>\n",
       "      <td>6.741935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-9</th>\n",
       "      <td>51.066667</td>\n",
       "      <td>28.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-10</th>\n",
       "      <td>36.741935</td>\n",
       "      <td>22.612903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-11</th>\n",
       "      <td>13.233333</td>\n",
       "      <td>6.733333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-12</th>\n",
       "      <td>1503.451613</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Sierra Leone</th>\n",
       "      <th>2014-8</th>\n",
       "      <td>24.741935</td>\n",
       "      <td>3.580645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-9</th>\n",
       "      <td>69.233333</td>\n",
       "      <td>4.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-10</th>\n",
       "      <td>112.193548</td>\n",
       "      <td>3.354839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-11</th>\n",
       "      <td>109.300000</td>\n",
       "      <td>0.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-12</th>\n",
       "      <td>21.903226</td>\n",
       "      <td>0.354839</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Daily average new cases  Daily average deaths\n",
       "Country      Month                                                 \n",
       "Guinea       2014-8                  4.161290              0.548387\n",
       "             2014-9                 10.466667              1.900000\n",
       "             2014-10                 1.096774              0.483871\n",
       "Liberia      2014-6                  1.333333              0.466667\n",
       "             2014-7                  3.032258              1.516129\n",
       "             2014-8                 10.806452              6.741935\n",
       "             2014-9                 51.066667             28.833333\n",
       "             2014-10                36.741935             22.612903\n",
       "             2014-11                13.233333              6.733333\n",
       "             2014-12              1503.451613              0.000000\n",
       "Sierra Leone 2014-8                 24.741935              3.580645\n",
       "             2014-9                 69.233333              4.333333\n",
       "             2014-10               112.193548              3.354839\n",
       "             2014-11               109.300000              0.400000\n",
       "             2014-12                21.903226              0.354839"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Write your answer here\n",
    "\n",
    "#Guinea\n",
    "\n",
    "guinea = pd.DataFrame()\n",
    "\n",
    "#we iterate in guinea_data folder and append every .csv file to our 'guinea' dataframe\n",
    "for file in glob.glob(DATA_FOLDER+\"/ebola/guinea_data/\"+'*.csv'):\n",
    "    eb = pd.read_csv(file, parse_dates=['Date']) # here we parse column 'Date' so we can handle dates easier\n",
    "    guinea = guinea.append(eb)\n",
    "\n",
    "#we set the index to 'Date' and 'Description' and we only need the 'Totals' column to copmute the daily averages\n",
    "guinea = (guinea.set_index(['Date','Description'])['Totals']).unstack('Description') #we make 'Description' values as column headers\n",
    "\n",
    "#we make a set of months so we can calculate the average\n",
    "months = set(guinea.index.month) #this way we get all months but the set deletes the duplicates\n",
    "new_cases = {}\n",
    "new_deaths = {}\n",
    "\n",
    "#now we iterate through our set of months and calculate the daily average of new cases and deaths\n",
    "for month in months: \n",
    "    #this is an easy way to get how many days has a particular month\n",
    "    days_in_month = (guinea.index.daysinmonth[guinea.index.month==month])[0] \n",
    "    \n",
    "    #here we look for all data from the same month and find the column named 'Total new cases registered so far'\n",
    "    #and just sum all values\n",
    "    month_sum_cases = (guinea[guinea.index.month==month])['Total new cases registered so far'].astype(float).sum()\n",
    "    \n",
    "    #for monthly deaths it's the same thing except we have in one file different column name, so we must handle both cases\n",
    "    month_sum_deaths1 = (guinea[guinea.index.month==month])['New deaths registered'].astype(float).sum()\n",
    "    month_sum_deaths2 = (guinea[guinea.index.month==month])['New deaths registered today'].astype(float).sum()\n",
    "    month_sum_deaths = np.nansum([month_sum_deaths1, month_sum_deaths2])\n",
    "\n",
    "    #calculate the daily average for both cases and deaths and append them to dict where the key is '2014'+month   \n",
    "    new_cases[\"2014-\"+str(month)] = month_sum_cases/days_in_month\n",
    "    new_deaths[\"2014-\"+str(month)] = month_sum_deaths/days_in_month\n",
    "    \n",
    "#make two new dataframes from dict, one for cases, one for deaths and merge them into one dataframe - 'guinea_final'\n",
    "guinea_new_cases = pd.DataFrame.from_dict(new_cases, orient='index')\n",
    "guinea_new_cases.index.name = 'Month'\n",
    "guinea_new_cases.columns = ['Daily average new cases']\n",
    "guinea_new_cases['Country'] = 'Guinea'\n",
    "\n",
    "guinea_deaths = pd.DataFrame.from_dict(new_deaths, orient='index')\n",
    "guinea_deaths.index.name = 'Month'\n",
    "guinea_deaths.columns = ['Daily average deaths']\n",
    "\n",
    "guinea_final = pd.merge(guinea_new_cases, guinea_deaths, left_index=True, right_index=True)\n",
    "guinea_final = guinea_final.set_index(['Country', guinea_final.index]) \n",
    "\n",
    "#for next countries we make almost everything the same, except for few things which will be noted\n",
    "\n",
    "#Liberia\n",
    "\n",
    "liberia = pd.DataFrame()\n",
    "\n",
    "for file in glob.glob(DATA_FOLDER+\"/ebola/liberia_data/\"+'*.csv'):\n",
    "    eb = pd.read_csv(file, parse_dates=['Date'])\n",
    "    liberia = liberia.append(eb)\n",
    "\n",
    "liberia = (liberia.set_index(['Date', 'Variable']))['National']\n",
    "liberia = pd.pivot_table(liberia.reset_index(), index='Date', columns='Variable', values='National')\n",
    "liberia = liberia.fillna(0)\n",
    "\n",
    "months = set(liberia.index.month)\n",
    "new_cases = {}\n",
    "new_deaths = {}\n",
    "for month in months:\n",
    "    days_in_month = (liberia.index.daysinmonth[liberia.index.month==month])[0]\n",
    "    \n",
    "    #here we had more expressions for new cases to handle\n",
    "    month_sum_cases1 = (liberia[liberia.index.month==month])['New Case/s (Suspected)'].astype(float).sum()\n",
    "    month_sum_cases2 = (liberia[liberia.index.month==month])['New Case/s (Probable)'].astype(float).sum()\n",
    "    month_sum_cases3 = (liberia[liberia.index.month==month])['New case/s (confirmed)'].astype(float).sum()\n",
    "    month_sum_cases = np.nansum([month_sum_cases1, month_sum_cases2, month_sum_cases3])\n",
    "    \n",
    "    month_sum_deaths = (liberia[liberia.index.month==month])['Newly reported deaths'].astype(float).sum()\n",
    "\n",
    "    new_cases[\"2014-\"+str(month)] = month_sum_cases/days_in_month\n",
    "    new_deaths[\"2014-\"+str(month)] = month_sum_deaths/days_in_month\n",
    "    \n",
    "liberia_new_cases = pd.DataFrame.from_dict(new_cases, orient='index')\n",
    "liberia_new_cases.index.name = 'Month'\n",
    "liberia_new_cases.columns = ['Daily average new cases']\n",
    "liberia_new_cases['Country'] = 'Liberia'\n",
    "\n",
    "liberia_deaths = pd.DataFrame.from_dict(new_deaths, orient='index')\n",
    "liberia_deaths.index.name = 'Month'\n",
    "liberia_deaths.columns = ['Daily average deaths']\n",
    "\n",
    "liberia_final = pd.merge(liberia_new_cases, liberia_deaths, left_index=True, right_index=True)\n",
    "liberia_final = liberia_final.set_index(['Country', liberia_final.index])\n",
    "\n",
    "#Sierra Leone\n",
    "\n",
    "sierra = pd.DataFrame()\n",
    "\n",
    "for file in glob.glob(DATA_FOLDER+\"/ebola/sl_data/\"+'*.csv'):\n",
    "    eb = pd.read_csv(file, na_values=['-'], parse_dates=['date'])\n",
    "    sierra = sierra.append(eb)\n",
    "    \n",
    "sierra_copy = sierra.copy()\n",
    "sierra_copy = sierra_copy.fillna(0)\n",
    "sierra = (sierra.set_index(['date','variable'])['National'])\n",
    "sierra = pd.pivot_table(sierra.reset_index(), index='date', columns='variable', values='National', aggfunc='first')\n",
    "sierra = sierra.fillna(0)\n",
    "\n",
    "#this one was the hardest because there are files that contain values in column 'National' which represent the sum of \n",
    "#new cases or deaths, but also there are files where we need manualy to sum the values because 'National' value is empty\n",
    "months = set(sierra.index.month)\n",
    "new_cases = {}\n",
    "new_deaths = {}\n",
    "for month in months:\n",
    "    days_in_month = (sierra.index.daysinmonth[sierra.index.month==month])[0]\n",
    "    \n",
    "    #first we get all 'National' values for column 'etc_new_deaths' in one month\n",
    "    month_sum_deaths = 0\n",
    "    month_deaths = (sierra[sierra.index.month==month])['etc_new_deaths'].astype(float)\n",
    "    for i in range(len(month_deaths)): #then we iterate for every value(every value represents one day or file)\n",
    "    #if the 'National' value for that file is '0', then then we sum the 'etc_new_deaths'\n",
    "    #for every other column(cities) and we get the sum for that day and add it to the monthly sum 'month_sum_deaths' \n",
    "        if month_deaths[i]==0: \n",
    "            temp = (sierra_copy[sierra_copy.date.dt.month==month])\n",
    "            month_sum_deaths += (temp.loc[temp['variable']=='etc_new_deaths'].drop(['date', 'variable', 'National'], axis=1).iloc[i].astype(float).sum())\n",
    "        #if the 'National' value is not '0', then we don't need to calculate anything,\n",
    "        #we just add the value for that day to the monthly sum\n",
    "        else: \n",
    "            month_sum_deaths += month_deaths[i]\n",
    "    \n",
    "    #the same thing goes with 'new cases', except we need to handle 4 different expressions\n",
    "    month_sum_cases = 0\n",
    "    month_cases1 = (sierra[sierra.index.month==month])['new_noncase'].astype(float)\n",
    "    for i in range(len(month_cases1)):\n",
    "        if month_cases1[i]==0:\n",
    "            temp = (sierra_copy[sierra_copy.date.dt.month==month])\n",
    "            month_sum_cases += (temp.loc[temp['variable']=='new_noncase'].drop(['date', 'variable', 'National'], axis=1).iloc[i].astype(float).sum())\n",
    "        else:\n",
    "            month_sum_cases += month_cases1[i]\n",
    "    month_cases2 = (sierra[sierra.index.month==month])['new_suspected'].astype(float)\n",
    "    for i in range(len(month_cases2)):\n",
    "        if month_cases2[i]==0:\n",
    "            temp = (sierra_copy[sierra_copy.date.dt.month==month])\n",
    "            month_sum_cases += (temp.loc[temp['variable']=='new_suspected'].drop(['date', 'variable', 'National'], axis=1).iloc[i].astype(float).sum())\n",
    "        else:\n",
    "            month_sum_cases += month_cases2[i]\n",
    "    month_cases3 = (sierra[sierra.index.month==month])['new_probable'].astype(float)\n",
    "    for i in range(len(month_cases3)):\n",
    "        if month_cases3[i]==0:\n",
    "            temp = (sierra_copy[sierra_copy.date.dt.month==month])\n",
    "            month_sum_cases += (temp.loc[temp['variable']=='new_probable'].drop(['date', 'variable', 'National'], axis=1).iloc[i].astype(float).sum())\n",
    "        else:\n",
    "            month_sum_cases += month_cases3[i]\n",
    "    month_cases4 = (sierra[sierra.index.month==month])['new_confirmed'].astype(float)\n",
    "    for i in range(len(month_cases4)):\n",
    "        if month_cases4[i]==0:\n",
    "            temp = (sierra_copy[sierra_copy.date.dt.month==month])\n",
    "            month_sum_cases += (temp.loc[temp['variable']=='new_confirmed'].drop(['date', 'variable', 'National'], axis=1).iloc[i].astype(float).sum())\n",
    "        else:\n",
    "            month_sum_cases += month_cases4[i]\n",
    "\n",
    "    new_cases[\"2014-\"+str(month)] = month_sum_cases/days_in_month\n",
    "    new_deaths[\"2014-\"+str(month)] = month_sum_deaths/days_in_month\n",
    "    \n",
    "sierra_new_cases = pd.DataFrame.from_dict(new_cases, orient='index')\n",
    "sierra_new_cases.index.name = 'Month'\n",
    "sierra_new_cases.columns = ['Daily average new cases']\n",
    "sierra_new_cases['Country'] = 'Sierra Leone'\n",
    "\n",
    "sierra_deaths = pd.DataFrame.from_dict(new_deaths, orient='index')\n",
    "sierra_deaths.index.name = 'Month'\n",
    "sierra_deaths.columns = ['Daily average deaths']\n",
    "\n",
    "sierra_final = pd.merge(sierra_new_cases, sierra_deaths, left_index=True, right_index=True)\n",
    "sierra_final = sierra_final.set_index(['Country', sierra_final.index])\n",
    "\n",
    "#at the end, we just create a new dataframe and append the data for every country\n",
    "final_frame = pd.DataFrame()\n",
    "final_frame = final_frame.append(guinea_final).append(liberia_final).append(sierra_final)\n",
    "final_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BARCODE</th>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">MID1</th>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methanosarcinaceae Methanimicrococcus</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methermicoccaceae Methermicoccus</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Ferroglobus</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Geoglobus</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloplanus</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloquadratum</th>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Halosimplex</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Natronobacterium</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Natronomonas</th>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Methanococci Methanococcales Methanocaldococcaceae Methanotorris</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Methanopyri Methanopyrales Methanopyraceae Methanopyrus</th>\n",
       "      <td>12</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Thermoplasmata Thermoplasmatales Thermoplasmatales_incertae_sedis Thermogymnomonas</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Acidothermaceae Acidothermus</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Beutenbergiaceae Salana</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Bogoriellaceae Bogoriella</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Cellulomonadaceae Tropheryma</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Corynebacteriaceae Turicella</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Cryptosporangiaceae Cryptosporangium</th>\n",
       "      <td>53</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermabacteraceae Dermabacter</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermabacteraceae Devriesea</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermatophilaceae Kineosphaera</th>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Frankineae_incertae_sedis Fodinicola</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Geodermatophilaceae Blastococcus</th>\n",
       "      <td>99</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Geodermatophilaceae Geodermatophilus</th>\n",
       "      <td>17</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Intrasporangiaceae Kribbia</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">MID9</th>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oceanospirillaceae Neptuniibacter</th>\n",
       "      <td>10</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oceanospirillales_incertae_sedis Spongiispira</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oleiphilaceae Oleiphilus</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pasteurellales Pasteurellaceae Haemophilus</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Moraxellaceae Alkanindiges</th>\n",
       "      <td>12</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Moraxellaceae Perlucidibaca</th>\n",
       "      <td>4</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azomonas</th>\n",
       "      <td>805</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azorhizophilus</th>\n",
       "      <td>862</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azotobacter</th>\n",
       "      <td>12</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Pseudomonas</th>\n",
       "      <td>3534</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Piscirickettsiaceae Cycloclasticus</th>\n",
       "      <td>4</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Piscirickettsiaceae Sulfurivirga</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Thiotrichaceae Beggiatoa</th>\n",
       "      <td>25</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Thiotrichales_incertae_sedis Fangia</th>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Sinobacteraceae Alkanibacter</th>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Sinobacteraceae Sinobacter</th>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Aquimonas</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Frateuria</th>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Ignatzschineria</th>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Luteibacter</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Synergistetes\" Synergistia Synergistales Synergistaceae Aminomonas</th>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Anaeroplasmatales Anaeroplasmataceae Asteroleplasma</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Haloplasmatales Haloplasmataceae Haloplasma</th>\n",
       "      <td>11</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Caldimicrobium</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Thermodesulfatator</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Geotoga</th>\n",
       "      <td>43</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Kosmotoga</th>\n",
       "      <td>16</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Opitutales Opitutaceae Opitutus</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Chlorarachniophyceae</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Streptophyta</th>\n",
       "      <td>10</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2396 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            Count  \\\n",
       "BARCODE Taxon                                                       \n",
       "MID1    Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      7   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      2   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...      3   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...      3   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      7   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      1   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      1   \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...      1   \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...      1   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...      1   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...      4   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...      1   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...      2   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...      4   \n",
       "        Archaea \"Euryarchaeota\" Methanococci Methanococ...      1   \n",
       "        Archaea \"Euryarchaeota\" Methanopyri Methanopyra...     12   \n",
       "        Archaea \"Euryarchaeota\" Thermoplasmata Thermopl...      2   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      2   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      1   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      1   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      1   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      2   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...     53   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      1   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      2   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      3   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      1   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...     99   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...     17   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      2   \n",
       "...                                                           ...   \n",
       "MID9    Bacteria \"Proteobacteria\" Gammaproteobacteria O...     10   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...      1   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...      1   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...      1   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...     12   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...      4   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    805   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    862   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...     12   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...   3534   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...      4   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...      1   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...     25   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...      2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      3   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      1   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      3   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      1   \n",
       "        Bacteria \"Synergistetes\" Synergistia Synergista...      3   \n",
       "        Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...      1   \n",
       "        Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...     11   \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...      1   \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...      1   \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...     43   \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...     16   \n",
       "        Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...      1   \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...      1   \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...     10   \n",
       "\n",
       "                                                                         GROUP  \\\n",
       "BARCODE Taxon                                                                    \n",
       "MID1    Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Methanococci Methanococ...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Methanopyri Methanopyra...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Thermoplasmata Thermopl...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "...                                                                        ...   \n",
       "MID9    Bacteria \"Proteobacteria\" Gammaproteobacteria O...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Synergistetes\" Synergistia Synergista...           Control 2   \n",
       "        Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...           Control 2   \n",
       "        Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...           Control 2   \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...           Control 2   \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...           Control 2   \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...           Control 2   \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...           Control 2   \n",
       "        Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...           Control 2   \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...           Control 2   \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...           Control 2   \n",
       "\n",
       "                                                             SAMPLE  \n",
       "BARCODE Taxon                                                        \n",
       "MID1    Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  unknown  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown  \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Methanococci Methanococ...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Methanopyri Methanopyra...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Thermoplasmata Thermopl...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "...                                                             ...  \n",
       "MID9    Bacteria \"Proteobacteria\" Gammaproteobacteria O...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Synergistetes\" Synergistia Synergista...    stool  \n",
       "        Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...    stool  \n",
       "        Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...    stool  \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...    stool  \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...    stool  \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...    stool  \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...    stool  \n",
       "        Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...    stool  \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...    stool  \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...    stool  \n",
       "\n",
       "[2396 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Write your answer here\n",
    "\n",
    "#first we initialize empty dataframe\n",
    "data = pd.DataFrame()\n",
    "\n",
    "#then we loop over 9 spreadsheets and append them to the empty dataframe\n",
    "for i in range(1,10):\n",
    "    mb = pd.read_excel(DATA_FOLDER+\"/microbiome/MID\"+str(i)+\".xls\", sheetname=\"Sheet 1\", header=None)\n",
    "    mb.columns = ['Taxon', 'Count'] #we name the columns\n",
    "    mb['BARCODE']=\"MID\"+str(i) #and add a nem column to know from which spreadsheet is the data \n",
    "    data = data.append(mb)\n",
    "    \n",
    "#we set the index to 'BARCODE' and 'Taxon' so that we have unique index\n",
    "data = data.set_index(['BARCODE', 'Taxon'])\n",
    "\n",
    "#just reading the metadata in separate dataframe and setting the index_col to 0 (that's the 'BARCODE' column)\n",
    "metadata = pd.read_excel(DATA_FOLDER+\"/microbiome/metadata.xls\", sheetname=\"Sheet1\", index_col=0)\n",
    "\n",
    "#we merge two dataframes based on index column from data and index column from metadata ('BARCODE')\n",
    "merged = pd.merge(data, metadata , left_index=True, right_index=True)\n",
    "\n",
    "#we can check if index is unique\n",
    "merged.index.is_unique\n",
    "\n",
    "#finally we replace the NaN values with 'unknown'\n",
    "merged = merged.fillna(\"unknown\")\n",
    "merged\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
